### DISCUSSION TITLE
Cannot connect to the locally deployed supabase using python client

### DISCUSSION THREAD

**Discussion by jais001** [supabase/supabase#30570] [Questions] [answered]
Hi,

Iam new to supabase. I have deployed supabase locally and wanted to connected to it using python. My python version 3.11.
I have followed the local setup document and installed the supabase locally in ubuntu 20.04
![Screenshot 2024-11-20 205513](https://github.com/user-attachments/assets/c819bf56-ff92-48da-8f61-17559e2c247d)

I am able to access my supabase dashboard and also was able to create table and insert data into it using the dashboard.

I wanted to connect to the supabase using python.
But I am not able to connect with it and I am getting error. I have tried the 
supabase_url: http://localhost:54323
supabase key: my anon key

but now I am getting connection refused error:

![1](https://github.com/user-attachments/assets/8e8a253c-8daf-4068-914c-8204bf386ef2)

I also tried with  http://localhost:54321 and  http://localhost:8000, when connecting with 8000 I am getting the value for that query as empty list. But i  have inserted data into it.

can someone help me resolve this issue. Iam not able to figure out the problem. thanks in advance

**Comment by GaryAustin1** [comment]
Seems like 8000 is the URL to use especially if you are not getting an error.
A return of empty array usually means you have RLS enabled and no policies or don't meet policies.  RLS is silent for selects.

**Reply by jais001** [reply]
Yeah, I have seen this from some post and did turn of RLS from dashboard of tables...but still not able to query using python..

**Reply by GaryAustin1** [reply]
Are you selfhosting or doing localdev?
What port are you using for accessing studio?  That should be the same for the supabase REST client.
Not sure have anything else for you, but 8000 not getting an error sounds like the right port for some supabase instance.

**Reply by jais001** [reply]
Iam self hosting. Iam accessing the dashboard with 8000

**Reply by GaryAustin1** [reply]
Then 8000 is the correct connection and you need to debug your select and triple check you have RLS disabled on the all lowercase student table.
Also make sure there is not an error (although with js client at least empty array would not return with an error).   
The only ways I know of for empty array are bad filter (you don't show one), do not meet RLS, or going to a different table/database than you expect.
I don't know the python client, but I don't think it caches (if you had RLS on then turned it off).  You could check this by change * to a column name to bust a cache if there is one from something else in your setup.

**Reply by jais001** [reply]
I have tried by using column name, still getting an empty list

**Reply by GaryAustin1** [reply]
My last post has the only reasons I can think of for an empty response (assuming that is the same as empty array in js... I don't know Python).

Can you show an image of your table in the table UI including the RLS status and the table name?

Do you have any other supabase instance with this table name in it?

**Reply by GaryAustin1** [reply]
Oh, and did you log the response error to insure it is "null" or no error code.

If you have logging working you could also look at the Gateway API log to see if the request is coming in to the instance.

**Reply by jais001** [reply]
Iam now able to get data when using service key. Still now iam not able to create any tables using the python code:
```
create_teacher_table = """
    CREATE TABLE IF NOT EXISTS teachers (
        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        name TEXT NOT NULL,
        subject_id UUID REFERENCES subjects (id) ON DELETE SET NULL,
        hired_at DATE NOT NULL,
        created_at TIMESTAMP DEFAULT now()
    );
    """
supabase.postgrest.rpc("pg_execute", {"sql": create_teacher_table})
```

No error are being returned, this table is not getting created

**Reply by GaryAustin1** [reply, marked as answer]
If you are getting data with the service_role key then you have RLS enabled on that table...

It is very dangerous and not really pattern to create tables with API calls.  If you insist on doing that then you need have a Postgres function  you call with rpc.  That function would then have to execute plpgsql code with SQL embedded in it to create your table.  

This is far from your first question and you should learn about Postgres functions, plpgsql, and how to execute SQL code in them.  Passing in SQL code though is highly risky.  This function would also have to be 'security definer' type to run.  You would need to protect it so regular users could not run it.  All of this is complex and beyond your current topic.  

I personally will not help anyone run SQL code generated from the API as I think that is something that requires detailed understanding of the code and risks involved.  Other users may be willing to help you do that.

Note with Python you can also use a Postgres library connected to the database direct port and execute SQL directly if you desire without using the REST API.

So, backing up to your first topic, you never showed your table and your RLS being disabled on it.  If RLS is enabled you have to have policies for it to work, unless you use service_role.  As service_role works, you pretty clearly have RLS enabled and no policies or don't meet them.

Please read thru the RLS guides as you are doing something wrong there... https://supabase.com/docs/guides/database/postgres/row-level-security
